cpp/src/arrow/io/file.cc

```diff
@@ -510,13 +511,19 @@ class MemoryMappedFile::MemoryMap : public MutableBuffer {
 
  private:
   // Initialize the mmap and set size, capacity and the data pointers
-  Status InitMMap(int64_t initial_size, bool resize_file = false) {
+  Status InitMMap(int64_t initial_size, bool resize_file = false,
+                  const int64_t offset = 0) {
     if (resize_file) {
       RETURN_NOT_OK(internal::FileTruncate(file_->fd(), initial_size));
     }
+    if (offset) {
+      if (offset % sysconf(_SC_PAGESIZE) != 0) {
```

Comment:
To be pedantic, `sysconf` might return -1 on some esoteric platforms and is very likely not going to work on windows?

```
static long page_size = sysconf(_SC_PAGESIZE);
if (page_size > 0 offset > 0 && offset % page_size != 0) {
  return Status::Invalid("Offset must be a multiple of the page size");
}
```

If this gets too hairy, you can just drop my recommendation and let mmap returns a failure.

Reply:
@fsaintjacques  Many thanks for the inputs! Indeed it seems heavy to implement the correct getpagesize under WIN32 and different platforms. I think it maybe better to let mmap() return the IOError if offset is not aligned. 
https://github.com/apache/arrow/blob/master/cpp/src/plasma/thirdparty/dlmalloc.c#L1565-L1608